source/agora/node/BlockSerialize.d

```diff
@@ -0,0 +1,81 @@
+/*******************************************************************************
+
+    Function definition and helper related to serialization
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.node.BlockSerialize;
+
+import agora.common.Data;
+import geod24.bitblob;
+
+/// Type of delegate SerializeDg
+public alias SerializeDg = void delegate(scope const(ubyte)[]) /*pure*/ nothrow @safe;
+
+public ubyte[] serializeFull (T) (scope const auto ref T record)
+    nothrow @trusted
+{
+    ubyte[] res;
+    scope SerializeDg dg = (scope const(ubyte []) data) nothrow @safe
+    {
+        res ~= data;
+    };
+
+    serializePart(record, dg);
+    return res;
+}
+
+/// Ditto
+public void serializePart (T) (scope const auto ref T record, scope SerializeDg dg)
+    /*pure*/ nothrow
+    if (is(T == struct))
+{
+    static assert(is(typeof(T.init.serialize(SerializeDg.init))),
+                "Struct `" ~ T.stringof ~
+                "` does not implement `serialize(scope SerializeDg) const nothrow` function");
+    record.serialize(dg);
+}
+
+/// Ditto
+public void serializePart () (scope const auto ref BitBlob!512LU record, scope SerializeDg dg)
+    /*pure*/ nothrow @safe
+{
+    dg(record[]);
+}
+
+/// Ditto
+public void serializePart (ubyte record, scope SerializeDg dg) /*pure*/ nothrow @trusted
```

Comment:
You might want to add overload for each basic types, because there are implicit conversions happening.

Reply:
^